File: Task1/generatedBill.js
function applyFlat10Discount(products, subTotal) {
  const discountAmount = 10;
  console.log("Flat 10 Discount Applied: ", discountAmount);
  return discountAmount;
}

function applyBulk5Discount(products, shortlistedProducts, subTotal) {
  let discountAmount = 0;

  for (const product of shortlistedProducts) {
    discountAmount += product[1] * product[2] * 0.05;
  }

  console.log("Bulk 5 Discount Applied: ", discountAmount);
  return discountAmount;
}

function applyBulk10Discount(products, subTotal) {
  const discountAmount = subTotal * 0.1;
  console.log("Bulk 10 Discount Applied: ", discountAmount);
  return discountAmount;
}

function applyTiered50Discount(products, shortlistedProducts, subTotal) {
  let discountAmount = 0;

  for (const product of shortlistedProducts) {
    discountAmount += product[1] * (product[2] - 15) * 0.5;
  }

  console.log("Tiered 50 Discount Applied: ", discountAmount);
  return discountAmount;
}

function generateBill(products, isGiftWrap) {
  let subTotal = 0;
  let subUnits = 0;
  const bulk5DiscountedProducts = [];
  const tiered50DiscountedProducts = [];

  const discounts = [false, false, false, false];
  let flagTiered = false;

  console.log("Product List:");
  console.log("Name Units Amount");

  for (const product of products) {
    subTotal += product[1] * product[2];
    subUnits += product[2];

    if (product[2] > 10) {
      bulk5DiscountedProducts.push(product);
      discounts[1] = true;
    }

    if (product[2] > 15) {
      tiered50DiscountedProducts.push(product);
      flagTiered = true;
    }

    console.log(product[0], product[2], product[1] * product[2]);
  }

  console.log("Subtotal: ", subTotal);

  if (subUnits > 30 && flagTiered) {
    discounts[3] = true;
  }

  if (subUnits > 20) {
    discounts[2] = true;
  }

  if (subTotal > 200) {
    discounts[0] = true;
  }

  if (discounts[0]) {
    subTotal -= applyFlat10Discount(products, subTotal);
  }

  if (discounts[1]) {
    subTotal -= applyBulk5Discount(products, bulk5DiscountedProducts, subTotal);
  }

  if (discounts[2]) {
    subTotal -= applyBulk10Discount(products, subTotal);
  }

  if (discounts[3]) {
    subTotal -= applyTiered50Discount(
      products,
      tiered50DiscountedProducts,
      subTotal
    );
  }

  const shippingFee = 10 * Math.ceil(subUnits / 10);
  subTotal += shippingFee;

  if (isGiftWrap) {
    subTotal += subUnits;
    console.log("Shipping and Gift Wrap Fee Applied: ", shippingFee + subUnits);
  } else {
    console.log("Shipping Fee Applied: ", shippingFee + subUnits);
  }

  console.log("Total Amount: ", subTotal);
}

const productsArray = [
  ["A", 20, 30],
  ["B", 40, 11],
  ["C", 50, 7],
];

generateBill(productsArray, true);

File: Task1/generated_bill.py
import math

def apply_flat_10_discount(products, sub_total):
    discount_amount = 10
    print("Flat 10 Discount Applied: ", discount_amount)
    return discount_amount

def apply_bulk_5_discount(products, shortlisted_products, sub_total):
    discount_amount = 0
    
    for product in shortlisted_products:
        discount_amount += (product[1] * product[2]) * 0.05
    
    print("Bulk 5 Discount Applied: ", discount_amount)
    return discount_amount

def apply_bulk_10_discount(products, sub_total):
    discount_amount = sub_total * 0.1
    print("Bulk 10 Discount Applied: ", discount_amount)
    return discount_amount

def apply_tiered_50_discount(products, shortlisted_products, sub_total):
    discount_amount = 0
    
    for product in shortlisted_products:
        discount_amount += (product[1] * (product[2] - 15)) * 0.5
    
    print("Tiered 50 Discount Applied: ", discount_amount)
    return discount_amount

def generate_bill(products, is_gift_wrap):
    sub_total = 0
    sub_units = 0
    bulk_5_discounted_products = []
    tiered_50_discounted_products = []
    
    discounts = [False] * 4
    flag_tiered = False
    
    print("Product List:")
    print("Name Units Amount")
    
    for product in products:
        sub_total += product[1] * product[2]
        sub_units += product[2]
        
        if product[2] > 10:
            bulk_5_discounted_products.append(product)
            discounts[1] = True
        
        if product[2] > 15:
            tiered_50_discounted_products.append(product)
            flag_tiered = True
        
        print(product[0], product[2], product[1] * product[2])
    
    print("Subtotal: ", sub_total)
            
    if sub_units > 30 and flag_tiered:
        discounts[3] = True
    
    if sub_units > 20:
        discounts[2] = True
    
    if sub_total > 200:
        discounts[0] = True
        
    
    if discounts[0]:
        sub_total -= apply_flat_10_discount(products, sub_total)
    
    if discounts[1]:
        sub_total -= apply_bulk_5_discount(products, bulk_5_discounted_products, sub_total)
    
    if discounts[2]:
        sub_total -= apply_bulk_10_discount(products, sub_total)
    
    if discounts[3]:
        sub_total -= apply_tiered_50_discount(products, tiered_50_discounted_products, sub_total)
     
    shipping_fee =  10 * (math.ceil(sub_units / 10))
    sub_total += shipping_fee
    
    if is_gift_wrap:
        sub_total += sub_units
        print("Shipping and Gift Wrap Fee Applied: ", shipping_fee + sub_units)
    else:
        print("Shipping Fee Applied: ", shipping_fee + sub_units)
    
    print("Total Amount: ", sub_total)

    

generate_bill([["A",20,30],["B",40,11],["C",50,7]],True)